### PR TITLE
Hive: Fix join issues when CBO is enabled

### PR DESCRIPTION
**Pull request by qphien**
After enabling CBO in Hive, there are some issues on MR when two iceberg table are joined. For example:
- Cannot find field from inspector
- ArrayIndexOutOfBoundsException when getting values from GenericRecord

These issues also happen with iceberg - non-iceberg table joins.

### CODE REVIEW

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java [hunk 1]
```diff
@@ -66,9 +67,7 @@ public void initialize(@Nullable Configuration configuration, Properties serDePr
     assertNotVectorizedTez(configuration);
 
     Schema tableSchema;
-    if (configuration.get(InputFormatConfig.TABLE_SCHEMA) != null) {
-      tableSchema = SchemaParser.fromJson(configuration.get(InputFormatConfig.TABLE_SCHEMA));
-    } else if (serDeProperties.get(InputFormatConfig.TABLE_SCHEMA) != null) {
+    if (serDeProperties.get(InputFormatConfig.TABLE_SCHEMA) != null) {
```
Comment (pvary) [inline mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:70]: Is this needed?
The original intent of the change was that we have the table schema at hand on the mappers/reducers. If we remove this then every mapper/reducer has to read the table once to get the schema.

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java [hunk 2]
```diff
@@ -82,7 +81,17 @@ public void initialize(@Nullable Configuration configuration, Properties serDePr
...
+    // in this case wrong recordStructField position leads wrong value or ArrayIndexOutOfBoundException
+    String[] distinctSelectedColumns = Arrays.stream(selectedColumns).distinct().toArray(String[]::new);
+    Schema projectedSchema = distinctSelectedColumns.length > 0 ?
+            tableSchema.select(distinctSelectedColumns) : tableSchema;
+    // the input split mapper handles does not belong to this table
+    // it is necessary to ensure projectedSchema equals to tableSchema,
+    // or we cannot find selectOperator's column from inspector
+    if (projectedSchema.columns().size() != distinctSelectedColumns.length) {
+      projectedSchema = tableSchema;
+    }
```
Comment (pvary) [inline mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:94]: @marton-bod: Could you please take a look? You know more about the schema projection.
Thanks,
Peter

File: mr/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandlerTestUtils.java [hunk 3]
```diff
@@ -60,18 +60,23 @@ static TestHiveShell shell() {
...
 
   static TestTables testTables(TestHiveShell shell, TestTables.TestTableType testTableType, TemporaryFolder temp)
-      throws IOException {
+          throws IOException {
 
     return testTableType.instance(shell.metastore().hiveConf(), temp);
   }
 
   static void init(TestHiveShell shell, TestTables testTables, TemporaryFolder temp, String engine) {
+    init(shell, testTables, temp, engine, "false");
```
Comment (pvary) [inline mr/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandlerTestUtils.java:69]: Could we use a boolean instead of the string?

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 4]
```diff
@@ -42,24 +42,41 @@
...
+import static org.apache.iceberg.types.Types.NestedField.optional;
 import static org.apache.iceberg.types.Types.NestedField.required;
 import static org.junit.runners.Parameterized.Parameter;
 import static org.junit.runners.Parameterized.Parameters;
 
 @RunWith(Parameterized.class)
 public class TestHiveIcebergStorageHandlerWithEngine {
 
-  private static final String[] EXECUTION_ENGINES = new String[] {"tez", "mr"};
+  private static final String[] EXECUTION_ENGINES = new String[]{"tez", "mr"};
```
Comment (pvary) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:53]: nit: I usually try to avoid formatting only changes

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 5]
```diff
@@ -42,24 +42,41 @@
...
+
+  private static final String[] CBO_ENABLES = new String[]{"true", "false"};
```
Comment (pvary) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:55]: We might not need this list as the values can not be changes. Could we just add this by hand at the `parameters()` method?

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 6]
```diff
@@ -107,6 +126,9 @@
   @Parameter(2)
   public TestTables.TestTableType testTableType;
 
+  @Parameter(3)
+  public String cboEnable;
```
Comment (pvary) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:130]: Boolean please

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 7]
```diff
@@ -142,29 +164,61 @@ public void testScanTable() throws IOException {
...
+            shell.executeStatement("SELECT first_name, customer_id FROM default.customers ORDER BY customer_id DESC");
 
     Assert.assertEquals(3, descRows.size());
-    Assert.assertArrayEquals(new Object[] {"Trudy", 2L}, descRows.get(0));
-    Assert.assertArrayEquals(new Object[] {"Bob", 1L}, descRows.get(1));
-    Assert.assertArrayEquals(new Object[] {"Alice", 0L}, descRows.get(2));
+    Assert.assertArrayEquals(new Object[]{"Trudy", 2L}, descRows.get(0));
+    Assert.assertArrayEquals(new Object[]{"Bob", 1L}, descRows.get(1));
+    Assert.assertArrayEquals(new Object[]{"Alice", 0L}, descRows.get(2));
+  }
```
Comment (pvary) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:173]: Am I right that these are formatting only changes?
It is much easier if we do not have them, so I usually try to avoid them in my PRs

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 8]
```diff
@@ -142,29 +164,61 @@ public void testScanTable() throws IOException {
...
+  @Test
+  public void testSelfJoin() throws IOException {
     testTables.createTable(shell, "orders", ORDER_SCHEMA, fileFormat, ORDER_RECORDS);
 
     List<Object[]> rows = shell.executeStatement(
-            "SELECT c.customer_id, c.first_name, o.order_id, o.total " +
-                    "FROM default.customers c JOIN default.orders o ON c.customer_id = o.customer_id " +
-                    "ORDER BY c.customer_id, o.order_id"
+            "SELECT o1.order_id, o1.customer_id, o1.total " +
+                    "FROM default.orders o1 JOIN default.orders o2 ON o1.order_id = o2.order_id ORDER BY o1.order_id"
```
Comment (pvary) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:215]: Why was this change needed?
Could you please help?
Thanks,
Peter

**Comment by pvary** [comment]
Thanks for the patch @qphien!
Really appreciate that you have taken the time to track this down!

Asked a few questions in the review comments. The general observations:
- Please do not do formatting only changes - These are making the review (and backport) harder
- I would recommend to use a boolean or Boolean instead of "false"/"true" strings
- There is one particular line I personally removed once accidentally and later realized that it is needed for performant queries (we might want to add a test case for it 😄). I think the line is still needed.
- I asked @marton-bod to review the projection related part since he was the one working on that.

Thanks,
Peter

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java (hunk 1 above)
Comment (qphien) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:70]: As show below:
* Mapper handles input split belongs to one table: 
https://github.com/apache/hive/blob/113f6af7528f016bf821f7a746bad496cc93f834/ql/src/java/org/apache/hadoop/hive/ql/io/HiveInputFormat.java#L406
* Function **copyTableJobPropertiesToConf** copies 'iceberg.mr.table.schema' property to jobConf:
https://github.com/apache/hive/blob/113f6af7528f016bf821f7a746bad496cc93f834/ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java#L2427-L2443
 
if we join two tables, only one iceberg schema exists in jobConf which leads wrong inspector:
* empty inspector(non-overlap in two table selected columns, e.g. 
`SELECT o.order_id, o.customer_id, o.total, p.name FROM default.orders o JOIN default.products p ON o.product_id = p.id ORDER BY o.order_id`
selected columns in table default.orders: [order_id, total, customer_id, product_id]
selected columns in table default.products: [id, name])
* incomplete inspector(overlap in two table selected columns, e.g.
`SELECT c.first_name, o.order_id FROM default.orders o JOIN default.customers c ON o.customer_id = c.customer_id ORDER BY o.order_id DESC`
selected columns in table default.customers: [first_name, customer_id]
selected columns in table default.orders: [order_id, customer_id]

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 4 above)
Comment (qphien) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:53]: Sorry, it's my fault, my IDEA automatically reformat code. I will pay more attention next time.

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 5 above)
Comment (qphien) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:55]: Okay, i will move CBO_ENABLES list to `parameters` method

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 8 above)
Comment (qphien) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:215]: As show in https://github.com/apache/hive/blob/113f6af7528f016bf821f7a746bad496cc93f834/ql/src/java/org/apache/hadoop/hive/ql/io/HiveInputFormat.java#L992-L998, 
 self join can lead duplicated columns in jobConf `hive.io.file.readcolumn.names`. 
In this case, it is necessary to test whether self join can work correctly.

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java (hunk 1 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:70]: Thanks for the explanation! This is something @marton-bod already tried his hands on. We might have to use the table identifier to prefix/postfix the schema....

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 8 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:215]: Oh.. I have missed that the FROM part is from the orders table.
Sorry

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java (hunk 1 above)
Comment (qphien) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:70]: Hi @marton-bod , is there any thread work on the proposal @pvary mentioned above? I can not find any related issue on github. 
Thanks.

**Comment by qphien** [comment]
I'm not sure whether these failure tests are related to this PR. I tested locally and all test cases were passed.
```
org.apache.iceberg.spark.extensions.TestCopyOnWriteDelete > testDeleteWithSerializableIsolation[catalogName = spark_catalog, implementation = org.apache.iceberg.spark.SparkSessionCatalog, config = {type=hive, default-namespace=default, clients=1, parquet-enabled=false, cache-enabled=false}, format = avro, vectorized = false] FAILED
    java.lang.RuntimeException: Failed to get table info from metastore default.table

        Caused by:
        org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)

            Caused by:
            java.net.SocketException: Broken pipe (Write failed)
```

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java (hunk 1 above)
Comment (marton-bod) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:70]: Hi @qphien . Indeed, we have had a similar problem before in https://github.com/apache/iceberg/issues/1708. 
My initial solution to fix it was to add the schema into the config object with a prefix based on the table identifier. For example, when joining together default.orders and default.customers, you'd have two properties in the config: `default.orders.iceberg.mr.table.schema` and `default.customers.iceberg.mr.table.schema`. This should allow you to add the schema for multiple tables without collisions/overwrites.

Eventually we ended up not adding in this fix, but instead just reverted the commit that caused the problem for us. However, this prefix solution can be one way to make things work.

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java (hunk 2 above)
Comment (marton-bod) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:94]: Looks good generally, but I wanted to clarify this comment:
`    // the input split mapper handles does not belong to this table
    // it is necessary to ensure projectedSchema equals to tableSchema,
    // or we cannot find selectOperator's column from inspector
`
Just for my understanding, can you give an example in what scenario we could face this issue where the Schema.select() gives back a different number of columns?

Comment (qphien) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:94]: With test case `testSelectedColumnsOverlapJoin`,  assuming that mapper is handling split belongs to table `default.orders`, columns set in `hive.io.file.readcolumn.names` are [order_id, customer_id], the inspector created for table `default.customers` just contains column [customer_id], when table `default.customers` selectOperator  is initialized, field `first_name` cannot be found from inspector we just created, so exception below is thrown 
```
cannot find field first_name from [org.apache.iceberg.mr.hive.serde.objectinspector.IcebergRecordObjectInspector$IcebergRecordStructField@a6807525]
	at org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorUtils.getStandardStructFieldRef
```

The cause of this exception is that the schema get from `Schema.select()` is not what we want, returning an inspector contains all table columns is an easy workaround to fix this issue.

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java (hunk 1 above)
Comment (qphien) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:70]: I think getting table schema from `serDeProperties` has the same effect as prefix solution. `iceberg.mr.table.schema` in `serDeProperties` is overlayed from corresponding tableDesc, so we can get correct table schema when `HiveIcebergSerde` is initialized.

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java (hunk 2 above)
Comment (marton-bod) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:94]: Thanks for the explanation!

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 9]
```diff
@@ -68,17 +83,20 @@
                   Types.TimestampType.withoutZone(), Types.StringType.get(), Types.BinaryType.get(),
                   Types.DecimalType.of(3, 1));
 
-  @Parameters(name = "fileFormat={0}, engine={1}, catalog={2}")
+  @Parameters(name = "fileFormat={0}, engine={1}, catalog={2}, cboEnable={3}")
   public static Collection<Object[]> parameters() {
     Collection<Object[]> testParams = new ArrayList<>();
     String javaVersion = System.getProperty("java.specification.version");
+    List<Boolean> cboEnables = ImmutableList.of(true, false);
```
Comment (marton-bod) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:90]: Instead of running all tests for both CBO on/off, can we turn it on just in those one or two unit test cases where we want to test it, by setting it via `shell.setHiveSessionValue()` at the beginning of the test? We're trying to prevent the number of test runs from exploding via the combination of ever-increasing test parameters.

Comment (qphien) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java]: @marton-bod OK, I have moved CBO test to some join test cases.

**Review by marton-bod — APPROVED** [review]
LGTM, thanks for picking this up @qphien

**Comment by rdblue** [comment]
Thanks for reviewing, @marton-bod and @pvary. Have all your concerns been addressed?

**Comment by marton-bod** [comment]
@rdblue yes, it should be good to go from my side. Thanks!

**Comment by pvary** [comment]
+1 from my side too

**Review by pvary — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks for fixing this, @qphien! And thanks for reviewing, @marton-bod and @pvary!
